compiler/packages/babel-plugin-react-compiler/src/Inference/InferMutationAliasingEffects.ts

```diff
@@ -1315,13 +1327,30 @@ class InferenceState {
     if (isRefOrRefValue(place.identifier)) {
       return 'mutate-ref';
     }
-    const kind = this.kind(place).kind;
+    const abstractValue = this.kind(place);
+    const kind = abstractValue.kind;
+
+    // Downgrade ShallowMutable to Mutable when mutated
+    if (kind === ValueKind.ShallowMutable) {
+      const values = this.values(place);
+      for (const value of values) {
+        const valueInfo = this.#values.get(value);
+        if (valueInfo && valueInfo.kind === ValueKind.ShallowMutable) {
+          this.#values.set(value, {
+            kind: ValueKind.Mutable,
+            reason: valueInfo.reason,
+          });
+        }
```

Comment:
This looks right. See my comment elsewhere about updating `mergeValueKinds()` to account for ShallowMutability. The logic there should basically be to treat ShallowMutable like Mutable when it joins with other things (if `Mutable | x => Mutable`, then `ShallowMutable | x => ShallowMutable`). Except `ShallowMutable | Mutable => Mutable`.

Reply:
The fact that you didn't hit the invariant in mergeValueKinds() also suggests we need some more tests. Something like

```js
function Component(props) {
  let x;
  if (props.cond) {
    x = [props.item]; // shallow mutable
  } else {
    x = []; // regular mutable
  }
  const z = x[0]; // z is mutable
}
```